service/storage_service.cc

```diff
@@ -1876,33 +1876,56 @@ future<std::map<gms::inet_address, float>> storage_service::effective_ownership(
+                [&ss, keyspace_name, final_ownership](const std::map<token, float>& token_ownership, std::unordered_map<sstring,
+                        std::unordered_set<gms::inet_address>>& datacenter_endpoints) {
+            return do_for_each(datacenter_endpoints, [&ss, &keyspace_name, final_ownership, &token_ownership, &datacenter_endpoints](std::pair<sstring,std::unordered_set<inet_address>>&& endpoints) {
+                return do_with(std::unordered_set<inet_address>(endpoints.second), [&ss, &keyspace_name, final_ownership, &token_ownership, &endpoints](const std::unordered_set<inet_address>& endpoints_map){
+                    return do_for_each(endpoints_map, [&ss, &keyspace_name, final_ownership, &token_ownership, &endpoints](const gms::inet_address& endpoint) {
+                        // calculate the ownership with replication and add the endpoint to the final ownership map
+                        try {
+                            return do_with(float(0.0f), dht::token_range_vector(ss.get_ranges_for_endpoint(keyspace_name, endpoint)),
+                                    [&ss, &keyspace_name, final_ownership, &token_ownership, &endpoint](float& ownership, const dht::token_range_vector& ranges) {
+                                ownership = 0.0f;
+                                return do_for_each(ranges, [&token_ownership,&ownership](const dht::token_range& r) mutable {
+                                    // get_ranges_for_endpoint will unwrap the first range.
+                                    // With t0 t1 t2 t3, the first range (t3,t0] will be splitted
+                                    // as (min,t0] and (t3,max]. Skippping the range (t3,max]
+                                    // we will get the correct ownership number as if the first
+                                    // range were not splitted.
+                                    if (!r.end()) {
+                                        return make_ready_future<>();
+                                    }
+                                    auto end_token = r.end()->value();
+                                    auto loc = token_ownership.find(end_token);
+                                    if (loc != token_ownership.end()) {
+                                        ownership += loc->second;
+                                    }
+                                    return make_ready_future<>();
+                                }).then([final_ownership = final_ownership, &ownership, &endpoint] {
+                                    (*final_ownership)[endpoint] = ownership;
+                                });
+                            });
+                        }  catch (...) {
+                            // In case ss.get_ranges_for_endpoint(keyspace_name, endpoint) is not found, just mark it as zero and continue
+                            (*final_ownership)[endpoint] = 0;
```

Comment:
Where did this come from? Just because you had an exception, doesn't mean it's okay to return zero!

Reply:
get_ranges_for_endpoint throws if there is no keyspace,endpoint match.
Which means that the specific endpoint has zero ownership.
Note that it only catches an exception that happens in get_ranges_for_endpoint.
I can catch only that specific exception to be on the safe side.

Reply:
`...` is not a specific exception. Any bad_alloc (or some other exception) here will be caught too.

Reply:
right, that's why I wrote that I'll change it to a specific exception so that only the case that there is no keyspace for the endpoint will be caught.

Reply:
Ah, I somehow missed "can"